src/realm/views.cpp

```diff
@@ -328,27 +363,54 @@ bool DescriptorOrdering::descriptor_is_sort(size_t index) const
 
 bool DescriptorOrdering::descriptor_is_distinct(size_t index) const
 {
-    return !descriptor_is_sort(index);
+    REALM_ASSERT(index < m_descriptors.size());
+    SortDescriptor* sort_descr = dynamic_cast<SortDescriptor*>(m_descriptors[index].get());
+    LimitDescriptor* limit_descr = dynamic_cast<LimitDescriptor*>(m_descriptors[index].get());
+    return (!sort_descr && !limit_descr); // dynamic cast of a sort descriptor to ColumnsDescriptor will succeed
+}
+
+bool DescriptorOrdering::descriptor_is_limit(size_t index) const
+{
+    REALM_ASSERT(index < m_descriptors.size());
+    LimitDescriptor* desc = dynamic_cast<LimitDescriptor*>(m_descriptors[index].get());
+    return (desc != nullptr);
 }
 
-const CommonDescriptor* DescriptorOrdering::operator[](size_t ndx) const
+const BaseDescriptor* DescriptorOrdering::operator[](size_t ndx) const
 {
     return m_descriptors.at(ndx).get(); // may throw std::out_of_range
 }
 
 bool DescriptorOrdering::will_apply_sort() const
 {
-    return std::any_of(m_descriptors.begin(), m_descriptors. end(), [](const std::unique_ptr<CommonDescriptor>& desc) {
+    return std::any_of(m_descriptors.begin(), m_descriptors.end(), [](const std::unique_ptr<BaseDescriptor>& desc) {
         REALM_ASSERT(desc.get()->is_valid());
         return dynamic_cast<SortDescriptor*>(desc.get()) != nullptr;
     });
 }
 
 bool DescriptorOrdering::will_apply_distinct() const
 {
-    return std::any_of(m_descriptors.begin(), m_descriptors.end(), [](const std::unique_ptr<CommonDescriptor>& desc) {
+    return std::any_of(m_descriptors.begin(), m_descriptors.end(), [](const std::unique_ptr<BaseDescriptor>& desc) {
+        REALM_ASSERT(desc.get()->is_valid());
+        return dynamic_cast<SortDescriptor*>(desc.get()) == nullptr && dynamic_cast<LimitDescriptor*>(desc.get()) == nullptr;
+    });
+}
+
+bool DescriptorOrdering::will_apply_limit() const
+{
+    return std::any_of(m_descriptors.begin(), m_descriptors.end(), [](const std::unique_ptr<BaseDescriptor>& desc) {
+        REALM_ASSERT(desc.get()->is_valid());
+        return dynamic_cast<LimitDescriptor*>(desc.get()) != nullptr;
```

Comment:
`dynamic_cast<>` is an expensive operation on some platforms, and merely checking the type can be done cheaply as `typeid(*desc) == typeid(LimitDescriptor)`.